### DISCUSSION TITLE
Got Revert on testing function : testFundUpdatesFundedDataStructure() in FundMe.t.sol

### DISCUSSION THREAD

**Discussion by Pradypta** [Cyfrin/foundry-full-course-cu#3034] [General]
**Question :** 
i've been following the tutorials at Foundry Fundamentals :14, and i got revert error when i tried to use this command : forge test --match-test testFundUpdatesFundedDataStructure -vvv

**Logs :**
Ran 1 test for test/FundMe.t.sol:FundMeTest
[FAIL: revert: You need to spend more ETH!] testFundUpdatesFundedDataStructure() (gas: 31870)
Traces:
  [31870] FundMeTest::testFundUpdatesFundedDataStructure()
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [14884] 0x90193C961A926261B756D1E5bb255e67ff9498A1::fund{value: 100000000000000000}()
    │   ├─ [8993] 0x34A1D3fff3958843C43aD80F30b94c510645C316::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200038 [2e5], 1, 1, 1
    │   └─ ← [Revert] revert: You need to spend more ETH!
    └─ ← [Revert] revert: You need to spend more ETH!

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 18.98ms (1.60ms CPU time)

Ran 1 test suite in 110.75ms (18.98ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/FundMe.t.sol:FundMeTest
[FAIL: revert: You need to spend more ETH!] testFundUpdatesFundedDataStructure() (gas: 31870)

Encountered a total of 1 failing tests, 0 tests succeeded

**Code :** 
**FundMe.sol :**
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

// Note: The AggregatorV3Interface might be at a different location than what was in the video!
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) private s_addressToAmountFunded;
    address[] private s_funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    AggregatorV3Interface private s_priceFeed;

    constructor(address priceFeed) {
        i_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }

    function getVersion() public view returns (uint256) {
        // AggregatorV3Interface priceFeed = AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        // return priceFeed.version();
        return s_priceFeed.version();
    }

    modifier onlyOwner() {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert NotOwner();
        _;
    }

    function withdraw() public onlyOwner {
        for (uint256 funderIndex = 0; funderIndex < s_funders.length; funderIndex++) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);

        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess,) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \
    //         yes  no
    //         /     \
    //    receive()?  fallback()
    //     /   \
    //   yes   no
    //  /        \
    //receive()  fallback()

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }
    // getter for testingk
    function getAddressToAmountFunded(address fundingAddress) external view returns(uint256){
        return s_addressToAmountFunded[fundingAddress];
    }
    
    function getFunder(uint256 index) external view returns(address){
        return s_funders[index];
    }
    function getPriceFeed() public view returns (AggregatorV3Interface) {
    return s_priceFeed;
}
}

// Concepts we didn't cover yet (will cover in later sections)
// 1. Enum
// 2. Events
// 3. Try / Catch
// 4. Function Selector
// 5. abi.encode / decode
// 6. Hash with keccak256
// 7. Yul / Assembly

**FundMe.t.sol :**
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity 0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/FundMe.s.sol";

contract FundMeTest is Test {
    FundMe fundMe;
    address USER = makeAddr("user");
    uint256 constant SEND_VALUE = 0.1 ether;
    uint256 constant STARTING_USER_BALANCE = 10 ether;

    function setUp() external {
        // fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();

        vm.deal(USER, STARTING_USER_BALANCE);
    }

    function testMnimumDollarIsFive() public {
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }

    function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), msg.sender);
    }

    // function testPriceFeedVersionIsAccurate() public {
    //     assertEq(fundMe.getVersion(), 4);
    // }
    function testPriceFeedVersionIsAccurate() public {
        if (block.chainid == 11155111) {
            uint256 version = fundMe.getVersion();
            assertEq(version, 4);
        } else if (block.chainid == 1) {
            uint256 version = fundMe.getVersion();
            assertEq(version, 6);
        }
    }

    function testFundFailsWithoutEnoughEth() public {
        vm.expectRevert(); // mengecek line selanjutnya akan error
        fundMe.fund(); //send 0 value
    }

    function testFundUpdatesFundedDataStructure() public {
        vm.prank(USER);
        fundMe.fund{value: SEND_VALUE}();

        uint256 amountFunded = fundMe.getAddressToAmountFunded(USER);
        assertEq(amountFunded, SEND_VALUE);
    }
}

**Comment by EngrPips** [comment]
Hello @Pradypta, The error is explicitly clear you just need to send more `ETH` 

```js
│ └─ ← [Revert] revert: You need to spend more ETH!
└─ ← [Revert] revert: You need to spend more ETH!
```

**Reply by Pradypta** [reply]
i know, the problem is :
i've been doing the exact thing as the tutorial and i send the same amount as the tutorials, and i've tried sending more eth by changing the SEND_VALUE to higher number but it still didn't work with the sam error tho. but 1 thing when i send more than the user balance it will revert as OutOfFund, i tried changing the value into slightly closer to the user balance and it didn't work as well

**Reply by EngrPips** [reply]
Show your `PriceConverter` contract.

**Reply by Pradypta** [reply]
there u go 
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";

// Why is this a library and not abstract?
// Why not an interface?
library PriceConverter {
    // We could make this public, but then we'd have to deploy it
    function getPrice(AggregatorV3Interface priceFeed) internal view returns (uint256) {
        // Sepolia ETH / USD Address
        // https://docs.chain.link/data-feeds/price-feeds/addresses
        // AggregatorV3Interface priceFeed = AggregatorV3Interface(
        //     0x694AA1769357215DE4FAC081bf1f309aDC325306
        // );
        (, int256 answer, , , ) = priceFeed.latestRoundData();
        // ETH/USD rate in 18 digit
        return uint256(answer * 10000000000);
    }

    // 1000000000
    function getConversionRate(
        uint256 ethAmount,
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {
        uint256 ethPrice = getPrice(priceFeed);
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        // the actual ETH/USD conversion rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }
}

**Reply by EngrPips** [reply]
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";

// Why is this a library and not abstract?
// Why not an interface?
library PriceConverter {
// We could make this public, but then we'd have to deploy it
function getPrice(AggregatorV3Interface priceFeed) internal view returns (uint256) {
// Sepolia ETH / USD Address
// https://docs.chain.link/data-feeds/price-feeds/addresses
// AggregatorV3Interface priceFeed = AggregatorV3Interface(
// 0x694AA1769357215DE4FAC081bf1f309aDC325306
// );
(, int256 answer, , , ) = priceFeed.latestRoundData();
// ETH/USD rate in 18 digit
return uint256(answer * 10000000000);
}

// 1000000000
function getConversionRate(
    uint256 ethAmount,
    AggregatorV3Interface priceFeed
) internal view returns (uint256) {
    uint256 ethPrice = getPrice(priceFeed);
    uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
    // the actual ETH/USD conversion rate, after adjusting the extra 0s.
    return ethAmountInUsd;
}
}
```

**Comment by KMean** [comment]
Here you are setting MINIMUM_USD to be 5 ether:
```solidity
uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
```
in your test you're setting SEND_VALUE to be 0.1 ether:
```solidity
uint256 constant SEND_VALUE = 0.1 ether;
```

if you want to send 0.1 change 
```solidity
uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
```
to
```solidity
uint256 public constant MINIMUM_USD = 1 * 10 ** 17;
```

or change your SEND_VALUE to be 5 ether;
```solidity
uint256 constant SEND_VALUE = 5 ether;
```

**Reply by Pradypta** [reply]
still reverting my friend, thanks tho

**Reply by KMean** [reply]
Yes sorry actually my answer was wrong as not considering the price conversion. 0.1 will be something as 200$ when converted using this as initial settings in your helper config:

```solidity
 MockV3Aggregator mockPriceFeed = new MockV3Aggregator(8, 2000e8);
```

so maybe ensure your mock initialization is correct in your HelperConfig.

**Comment by heheboii11** [comment]
I think you have not mocked the price converter for testing in the anvil. Did you use mock for ```priceconverter``` ?? If not please deploy the mock version of priceconverter. In the error you are getting you are getting the price of the assest is ```2e5``` but it should be a very high number, in the order of ```10e18```

**Comment by EngrPips** [comment]
Hello @Pradypta, Please show your `DeployFundMe` file.
